### PR TITLE
New table "diaspora-contact" (new paradigm) v34 FINAL

### PR DESCRIPTION
**Pull request by MrPetovan**
It turns out using data structures for parameter streamlines the migration to the new paradigm. The `WebFingerUri` class is small, but knowing what is in a parameter makes it straightforward. In particular, it ensures there's actually something in the parameter, or the issue would have been caught much earlier, when the parameter class was instantiated in the first place. The parameter object can then be passed along without having to perform `empty()` checks again.

### CODE REVIEW

**Comment by nupplaphil** [comment]
@MrPetovan  ... ready to merge or do you think @annando should review it again?

**Comment by MrPetovan** [comment]
I'm sorry, @annando definitely has the final word on this PR that he originally started.

**Comment by annando** [comment]
The weekend had been busy, I will try having a look at the stuff this evening.

File: database.sql
Comment (annando) [inline database.sql]: I just realized that we don't need the `baseurl` here. The `gsid` is enough. Instead we can add the `baseurl` to the view.

File: static/dbview.config.php
Comment (annando) [inline static/dbview.config.php]: Please change this to `"baseurl" => ["gserver", "url"],`

File: static/dbview.config.php
Comment (annando) [inline static/dbview.config.php]: Plase add
```
LEFT JOIN `gserver` ON `gserver`.`id` = `diaspora-contact`.`gsid`
```

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (annando) [inline src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: Do you at some point in time perform some `$gsid = GServer::getID(DiasporaContact->baseurl);` or is it needed here?

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (annando) [inline src/Protocol/Diaspora/Repository/DiasporaContact.php]: Do you need all these fields here?

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (annando) [inline src/Protocol/Diaspora/Repository/DiasporaContact.php:237]: Isn't this redundant to the functionality in the `save` function?

**Review by annando — CHANGES_REQUESTED** [review]
If nitpicking weren't yet invented, then I surely would be a good explanation of what "nitpicking" is ...

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:200]: You're right, I've been flip-flopping about where to put this and when I finally put it in the `save` method I didn't remove it from here.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php]: Yes, they're used inside `Contact::getByURL` to determine if the contact needs to be updated 🤷‍♂️

I guess I could add them manually in the method itself, it'd be better.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: No, it's done during the probe so we expect it to be set in the DiasporaContact object already.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:237]: Yes, but we still need it here to retrieve the follow/er/s counts, and this way we can set it in the entity object as well.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (annando) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php]: Okay. I wasn't sure if this was some kind of leftover.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (annando) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:237]: Could we pass the `uriid` down to `save` and then only calculate it when missing?

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (annando) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: I guess then we should add it here when not provided, just to be safe?

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:237]: It's already the case, see line 124.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: The table that you created includes a provision for an empty `gsid`, which is what I based myself on to create the classes. This means DiasporaContact can have an empty `gsid` property as well, so this doesn't warrant a check in `save` the same way we need to check for an `uri-id`.

Comment (annando) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: When there is a baseurl there will be a gsid - especially all known Diaspora implementations are detectable via GServer.

Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: So the `gsid` is mandatory for `diaspora-contact`?

Comment (annando) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: No. Neither `gsid` nor `baseurl` are mandatory. It's just that `baseurl` is superfluous when `gsid` is there. Both fields are in the "nice to have, but the world will turn without them as well" category.

Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php:124]: Got it.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php]: Never mind, these fields were already automatically added inside of `Contact::getByURL`.

**Comment by annando** [comment]
One final question: Have you had this running on your machine? Before merging I really would like to see that Diaspora messages still arrive and that the error log isn't exploding.

**Comment by MrPetovan** [comment]
Now that the review is complete, I will do that over the next couple of days.

**Comment by annando** [comment]
BTW: I just realized that you set the database version to some higher than needed number. Was this intended?

**Comment by MrPetovan** [comment]
I just separated the database structure changes in three different commits, so I updated the number each time so that the codebase would always be coherent even if someone checked out a single commit.

**Comment by MrPetovan** [comment]
This branch is now running on my production node and the `diaspora-contact` table is starting to fill. There was a slight hiccup where I expected a `created` field that I never added in the table structure, and then I had to figure out why it was still missing from the query results, and then I remembered I had to force recreate the views.

Now it's running smoothly.

**Comment by MrPetovan** [comment]
I just fixed a typo in a database field name that was preventing public keys from being saved in the table which was crashing the Diaspora message reception. I'll run the branch a little more.

**Review by annando — APPROVED** [review]

**Comment by MrPetovan** [comment]
So far the only eyebrow-raising log messages I'm getting are variations of this one:
```
2022-12-01T10:51:48Z worker [NOTICE]: No URL for contact {"id":1234567,"addr":"xxxxxx@libranet.de","batch mode":true,"batch":"","notify":"https://libranet.de/dfrn_notify/xxxxxx","worker_id":"10356ff","worker_cmd":"Delivery"} - {"file":"Diaspora.php","line":2999,"function":"transmit","uid":"689338","process_id":9765}
```
And this one:
```
2022-12-01T10:54:21Z index [NOTICE]: Fetching diaspora key {"handle":"xxxxxx@friendica.xyz","callstack":"Diaspora::decodeRaw, Receive::decodePost, Receive::receivePublic, Receive::post, BaseModule::run, App::runFrontend"} - {"file":"Diaspora.php","line":795,"function":"key","uid":"2f3fcb","process_id":10056}
2022-12-01T10:54:22Z index [NOTICE]: Couldn't get a key for handle xxxxxx@friendica.xyz. Discarding. [] - {"file":"Diaspora.php","line":314,"function":"decodeRaw","uid":"2f3fcb","process_id":10056}
```
Both are Friendica contacts, is it expected they get processed (and fail) in the Diaspora transmission/reception?

**Comment by MrPetovan** [comment]
I added some more information to the logging for the first message:
```
2022-12-01T11:51:49Z worker [NOTICE]: No URL for contact {"id":493240,"addr":"xxxxxx@squeet.me","batch mode":true,"batch":"","notify":"https://squeet.me/dfrn_notify/xxxxxx","network":"dspr","worker_id":"2c06fad","worker_cmd":"Delivery"} - {"file":"Diaspora.php","line":2999,"function":"transmit","uid":"994ba9","process_id":22772}
```
`"network":"dspr"` for a Friendica contact?

**Comment by annando** [comment]
Yeah, Friendica can be recognized as Diaspora as well.

**Comment by MrPetovan** [comment]
Then we're all good.

**Comment by annando** [comment]
When your system now works fine with this branch, then I delegate the task to check and merge it to @nupplaphil since he much better has got an overview about this whole class thingy.

**Comment by MrPetovan** [comment]
it's been fine for a couple days, I've been able to interact with Diaspora contacts (who suddenly started receiving my posts again 😅)

File: src/Protocol/Diaspora.php
Comment (nupplaphil) [inline src/Protocol/Diaspora.php:799]: No logging ?

File: src/Protocol/Diaspora.php
Comment (nupplaphil) [inline src/Protocol/Diaspora.php:2361]: Logging?

File: src/Protocol/Diaspora.php
Comment (nupplaphil) [inline src/Protocol/Diaspora.php:2385]: Logging?

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (nupplaphil) [inline src/Protocol/Diaspora/Repository/DiasporaContact.php]: This seems a little bit inperformant, selecting the whole entity  just for one field
--> shouldn't we use here a direct table call instead?
We're already in the repository, so it won't break the pattern doing so

**Review by nupplaphil — CHANGES_REQUESTED** [review]
just minor things

**Comment by nupplaphil** [comment]
conflicts :D

File: src/Protocol/Diaspora.php
Comment (MrPetovan) [reply src/Protocol/Diaspora.php:799]: There was none before, not sure it would be helpful.

File: src/Protocol/Diaspora.php
Comment (MrPetovan) [reply src/Protocol/Diaspora.php:2361]: There was none before, not sure it would be helpful.

File: src/Protocol/Diaspora.php
Comment (MrPetovan) [reply src/Protocol/Diaspora.php:2385]: There was none before, not sure it would be helpful.

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php]: Good point, sounds good.

**Comment by MrPetovan** [comment]
Of course, people keep changing the database schema! 👀

**Comment by annando** [comment]
I just wanted to fill up the database version numbers in between :grin:

File: src/Protocol/Diaspora/Repository/DiasporaContact.php
Comment (MrPetovan) [reply src/Protocol/Diaspora/Repository/DiasporaContact.php]: Simplified.

**Comment by MrPetovan** [comment]
Database version 1500! 🎉

**Comment by annando** [comment]
But only because you skipped some versions :-)

**Comment by MrPetovan** [comment]
I made it easier to review!

**Comment by MrPetovan** [comment]
Yayyy!

**Pull request merged**
